aesd-char-driver/main.c

```diff
@@ -0,0 +1,222 @@
+/**
+ * @file aesdchar.c
+ * @brief Functions and data related to the AESD char driver implementation
+ *
+ * Based on the implementation of the "scull" device driver, found in
+ * Linux Device Drivers example code.
+ *
+ * @author Dan Walkes
+ * @date 2019-10-22
+ * @copyright Copyright (c) 2019
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/init.h>
+#include <linux/printk.h>
+#include <linux/types.h>
+#include <linux/cdev.h>
+#include <linux/fs.h> // file_operations
+#include <linux/slab.h>
+#include <linux/uaccess.h>
+
+#include "aesdchar.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Chandana Challa");
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+
+int aesd_open(struct inode *inode, struct file *filp)
+{
+    struct aesd_dev *dev = NULL;
+    PDEBUG("open");
+
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp)
+{
+    PDEBUG("release");
+    filp->private_data = NULL;
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count,
+                loff_t *f_pos)
+{
+    ssize_t retval = 0;
+    size_t entry_offset = 0;
+    struct aesd_buffer_entry *entry = NULL;
+    ssize_t read_bytes = 0;
+
+    
+    if ( (NULL == filp) || (NULL == buf))
+    {
+        PDEBUG("ERROR: aesd_read invalid arguments");
+        return -EINVAL;
+    }
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+
+    entry = aesd_circular_buffer_find_entry_offset_for_fpos(&aesd_device.buffer,
```

Comment:
@chchbuff are you guaranteed aesd_device.buffer won't be modified for this function call?  Are you guaranteed `entry` won't be modified during handling below?